### PR TITLE
Release 2.16.0

### PR DESCRIPTION
**Pull request by EvansPie**
Release 2.16.0

- Integrate iPay88
- Add Headless analytics events
- Add properties in analytics events
- Return error on Apple Pay on simulator

### CODE REVIEW

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/ApplePayTokenizationViewModel.swift [hunk 1]
```diff
@@ -397,6 +397,24 @@ extension ApplePayTokenizationViewModel: PKPaymentAuthorizationViewControllerDel
         didAuthorizePayment payment: PKPayment,
         handler completion: @escaping (PKPaymentAuthorizationResult) -> Void
     ) {
+#if targetEnvironment(simulator)
+        if payment.token.paymentData.count == 0 {
+            let err = PrimerError.invalidArchitecture(
+                description: "Apple Pay cannot return tokenization data in the simulator",
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/ApplePayTokenizationViewModel.swift:403]: Merchants don't know what "tokenization data" means here.
I'd put sth like: "Apple Pay does not work with Primer when used in the simulator due to a limitation from Apple Pay."

File: Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckout.swift [hunk 2]
```diff
@@ -39,7 +47,28 @@ public class PrimerHeadlessUniversalCheckout {
...
+        
+        self.checkoutSessionId = UUID().uuidString
+        self.timingEventId = UUID().uuidString
+        
+        let sdkEvent = Analytics.Event(
+            eventType: .sdkEvent,
+            properties: SDKEventProperties(
+                name: #function,
+                params: [
+                    "intent": PrimerInternal.shared.intent?.rawValue ?? "null"
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckout.swift:61]: Any way we can pass al parameters? `clientToken` and `settings` are also valuable?

File: Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckout.swift [hunk 3]
```diff
@@ -181,16 +231,43 @@ public class PrimerHeadlessUniversalCheckout {
     }
     
     public static func makeButton(for paymentMethodType: String) -> UIButton? {
+        let sdkEvent = Analytics.Event(
+            eventType: .sdkEvent,
+            properties: SDKEventProperties(
+                name: #function,
+                params: nil))
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckout.swift:238]: Can we have full visibility on the args used?

File: Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckout.swift [hunk 4]
```diff
@@ -39,7 +47,28 @@ public class PrimerHeadlessUniversalCheckout {
...
+                ]))
+        
+        let timingStartEvent = Analytics.Event(
+            eventType: .timerEvent,
+            properties: TimerEventProperties(
+                momentType: .start,
+                id: self.timingEventId))
+        
+        Analytics.Service.record(events: [sdkEvent, timingStartEvent])
+        Analytics.Service.sync()
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckout.swift:71]: Why do you need to call `sync` all the time?

File: Sources/PrimerSDK/Classes/Data Models/ClientToken.swift [hunk 5]
```diff
@@ -28,6 +28,10 @@ struct DecodedJWTToken: Codable {
     var qrCode: String?
     var accountNumber: String?
     
+    // iPay88
+    var backendCallbackUrl: String?
+    var primerTransactionId: String?
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/Data Models/ClientToken.swift:33]: Is there any plan to split this file into multiple files? It seems that having only one file for all possibilities of the client token is a bit hard to follow 🤔

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+    private var didCancel: (() -> Void)?
+    private var didComplete: (() -> Void)?
+    private var didFail: ((_ err: PrimerError) -> Void)?
+#endif
+    
+    private lazy var iPay88NumberFormatter: NumberFormatter = {
+        let numberFormatter = NumberFormatter()
+        numberFormatter.numberStyle = .decimal
+        numberFormatter.decimalSeparator = "."
+        numberFormatter.groupingSeparator = ","
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:39]: Are we 100% sure this is not tied to any locale?

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,547 @@
...
+        numberFormatter.maximumFractionDigits = 2
+        numberFormatter.minimumFractionDigits = 2
+        return numberFormatter
+    }()
+    
+    override func validate() throws {
+        guard let decodedJWTToken = PrimerAPIConfigurationModule.decodedJWTToken, decodedJWTToken.isValid else {
+            let err = PrimerError.invalidClientToken(
+                userInfo: ["file": #file, "class": "\(Self.self)", "function": #function, "line": "\(#line)"],
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:48]: I'm seeing you using this everywhere. I understand you cannot use another function cuz `#file` and `#function` would point to another place. 
But could use preprocessing?

```
#define ANALYTICS_DEFAULT_USER_INFO ["file": #file, "class": "\(Self.self)", "function": #function, "line": "\(#line)"]
```

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,547 @@
...
+        
+        if AppState.current.currency != .MYR {
+            let err = PrimerError.invalidClientSessionValue(
+                name: "currency",
+                value: AppState.current.currency?.rawValue,
+                userInfo: ["file": #file, "class": "\(Self.self)", "function": #function, "line": "\(#line)"],
+                diagnosticsId: nil)
+            ErrorHandler.handle(error: err)
+            errors.append(err)
+        }
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:104]: These checks should be done before showing the payment method button to the user / passing it to `onAvailablePaymentMethodsLoad`.

No need to show a payment method if know it won't work

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,547 @@
...
+            ErrorHandler.handle(error: err)
+            throw err
+        }
+    }
+    
+    override func performPreTokenizationSteps() -> Promise<Void> {
+        let event = Analytics.Event(
+            eventType: .ui,
+            properties: UIEventProperties(
+                action: .click,
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:219]: Don't think that's the right place to call this.
We would be logging a `.click` event in headless checkout, which doesn't make much sense here 🤔

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,547 @@
...
+                seal.fulfill()
+            }
+            .ensure {
+
+            }
+            .catch { err in
+                seal.reject(err)
+            }
+            
+#else
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:255]: Same thing here: we shouldn't return `iPay88` in the list of payment methods if the SDK is missing

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift [hunk 11]
```diff
@@ -0,0 +1,547 @@
...
+                    self.primerIPay88ViewController.isModalInPresentation = true
+                    self.primerIPay88ViewController.modalPresentationStyle = .fullScreen
+                }
+
+                self.willPresentPaymentMethodUI?()
+                PrimerUIManager.primerRootViewController?.present(self.primerIPay88ViewController, animated: true, completion: {
+                    DispatchQueue.main.async {
+                        PrimerHeadlessUniversalCheckout.current.delegate?.primerHeadlessUniversalCheckoutPaymentMethodDidShow?(for: self.config.type)
+                        self.didPresentPaymentMethodUI?()
+                        seal.fulfill(())
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:449]: What does `fulfill(())` means? (empty parathensis)

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift [hunk 12]
```diff
@@ -0,0 +1,547 @@
...
+            self.didCancel = {
+                let err = PrimerError.cancelled(
+                    paymentMethodType: self.config.type,
+                    userInfo: ["file": #file, "class": "\(Self.self)", "function": #function, "line": "\(#line)"],
+                    diagnosticsId: nil)
+                ErrorHandler.handle(error: err)
+                pollingModule.cancel(withError: err)
+            }
+            
+            self.didFail = { err in
```
Comment (xevious78) [inline Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:477]: If I get this right: this automatically outputs a log to Primer?

**Review by xevious78 — APPROVED** [review]
Looking good. Got some comments

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/ApplePayTokenizationViewModel.swift (hunk 1 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/ApplePayTokenizationViewModel.swift]: Fixed

File: Sources/PrimerSDK/Classes/Data Models/ClientToken.swift (hunk 5 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/Data Models/ClientToken.swift:33]: Yeah, there is once the `clientToken` is refactored.

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 8 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:104]: Isn't this supposed to be driven by the BE? If a PM is not valid for a PM, then BE shouldn't return it in the response, and therefore not rendered. Right?

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 9 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:219]: Yeah, you're right

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 11 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:449]: Same as fulfilling with `Void`

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 10 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:255]: `iPay88` would never be here, because the view model will be removed from the `PrimerAPIConfiguration.paymentMethodConfigViewModels` if `PrimerIPay88SDK` is missing.

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 6 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:39]: `iPay88` has no docs on this, I tried with a couple of combinations. Never managed to get the `groupingSeparator` cause they don't allow payments in sandbox over 999.99, they throw an `Transaction overlimit` error. @semirp anything on Android?

File: Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckout.swift (hunk 4 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/Core/Checkout Components/PrimerHeadlessUniversalCheckout.swift]: No need actually. Removed them.

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 7 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:48]: +1

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 12 above)
Comment (EvansPie) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:477]: If it cancels on HUC, an `cancellation` error will be returned. If it's dropin, it will return on the dropin UI.

If it's an error, it will be passed via the delegates.

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 6 above)
Comment (semirp) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:39]: They just say we should use 1,234.55 format @EvansPie

File: Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift (hunk 8 above)
Comment (semirp) [reply Sources/PrimerSDK/Classes/User Interface/TokenizationViewModels/IPay88TokenizationViewModel.swift:104]: @EvansPie similar to what we did to other APMs. We verify it **before** and log warnings if something is wrong. Probably you did it for Apple Pay
